yo1k/tic_tac_toe/core/game.py

```diff
@@ -0,0 +1,296 @@
+        self.__mark = mark
+        self.__wins = 0
+
+    def mark(self) -> Mark:
+        return self.__mark
+
+    @property
+    def wins(self) -> int:
+        return self.__wins
+
+    @wins.setter
+    def wins(self, wins: int):
+        self.__wins = wins
+
+    def __str__(self):
+        return f"{type(self)}(" \
+               f"mark={self.__mark}," \
+               f"wins={self.__wins})"
+
+
+class Phase(Enum):
+    ONSET = auto()
+    INROUND = auto()
+    OUTROUND = auto()
+
+
+class Cell:
+    """A game board cell."""
+
+    def __init__(self, x: int, y: int):
+        assert 0 <= x < Board.size(), f"{x}"
+        assert 0 <= y < Board.size(), f"{y}"
+        self.__x = x
+        self.__y = y
+
+    @property
+    def x(self) -> int:
+        return self.__x
+
+    @property
+    def y(self) -> int:
+        return self.__y
+
+    def __str__(self):
+        return f"{type(self)}(" \
+               f"x={self.__x}," \
+               f"y={self.__y})"
+
+
+class Board:
+    def __init__(self, cells: Sequence[list[Optional[Mark]]] = None):
+        self.__cells = Board.__empty_cells() if cells is None else cells
+
+    def _set(self, cell: Cell, mark: Mark):
+        self.__cells[cell.x][cell.y] = mark
+
+    def _get(self, cell: Cell) -> Optional[Mark]:
+        return self.__cells[cell.x][cell.y]
+
+    def _clean(self):
+        self.__cells = Board.__empty_cells()
+
+    @property
+    def cells(self):
+        return self.__cells
+
+    @staticmethod
+    def size() -> int:
+        return 3
+
+    @staticmethod
+    def __empty_cells() -> list[list[Optional[Mark]]]:
+        return [[None for _ in range(Board.size())] for _ in range(Board.size())]
+
+    def __str__(self):
+        return f"{type(self)}(" \
+               f"cells={self.__cells},"
+
+
+class State:
+    """Full game state which is enough to restore a paused game."""
+
+    def __init__(self,
+                 game_rounds: int,
+                 players: Sequence[Player],
+                 board: Board,
+                 phase=Phase.ONSET,
+                 cur_round: int = 0,
+                 step: int = 0):
+        self.game_rounds = game_rounds
+        self.round = cur_round
+        self.players = players
+        self.step = step
+        self.phase = phase
+        self.board = board
+
+    def __str__(self):
+        return f"{type(self)}(" \
+               f"game_rounds={self.game_rounds}," \
+               f"round={self.round}," \
+               f"players={self.players}," \
+               f"step={self.step}," \
+               f"phase={self.phase}," \
+               f"board={self.board})"
+
+
+class Action:
+    def __init__(self, surrender: bool, occupy: Optional[Cell], next_round: bool):
+        assert (surrender is True and occupy is None and next_round is False) or \
+               (surrender is False and occupy is not None and next_round is False) or \
+               (surrender is False and occupy is None and next_round is True)
```

Comment:
It'd be good to have an error message specified so that it's clear what were the invalid arguments.